process_report/invoices/pi_specific_invoice.py

```diff
@@ -37,30 +61,134 @@ class PIInvoice(invoice.Invoice):
-            pi_instituition = pi_projects[invoice.INSTITUTION_FIELD].iat[0]
-            pi_projects.to_csv(
-                f"{self.name}/{pi_instituition}_{pi} {self.invoice_month}.csv"
+        def _create_html_invoice():
+            environment = Environment(loader=FileSystemLoader(TEMPLATE_DIR_PATH))
+            template = environment.get_template("pi_invoice.html")
+            content = template.render(
+                data=pi_dataframe,
             )
 
+            with tempfile.NamedTemporaryFile(
+                mode="w", delete=False, suffix=".html"
+            ) as f:
+                f.write(content)
+                return f.name
+
+        def _create_pdf_invoice():
+            driver = self._get_chrome_driver()
+            driver.get("file://" + invoice_html_path)
+
+            """
+            To ensure the HTML invoice table always fit the page when
+            printed to PDF, certain things are done below. Before my
+            explanation, some context is needed:
+
+            When printing to pdf, the Chrome driver picks the default
+            page size of A4, which is 8.27in x 11.69in. In CSS, the length
+            units include pixels and inches, and it seems 1 inch
+            equals 96 pixels. This explains the constants below.
+
+            In order to ensure the invoice table fits along the width
+            (since we’re printing landscape) of the A4 page, we want
+            to scale the table down.
+
+            This scaling value should be =  width of A4 / length of table.
+            """
+            A4_WIDTH_INCHES = 11.69
+            PX_PER_INCH = 96
+            invoice_table = driver.find_element(by=By.TAG_NAME, value="table")
+            px_width = float(
+                invoice_table.value_of_css_property("width").split("px")[0]
+            )
+            scale_value = A4_WIDTH_INCHES / (px_width / PX_PER_INCH)
+
+            print_options = PrintOptions()
+            print_options.orientation = "landscape"
+            print_options.scale = scale_value
+            print_options.shrink_to_fit = True
+
+            pdf_str = driver.print_page(print_options=print_options)
+            pdf = base64.b64decode(pdf_str)
+            with open(
+                f"{self.name}/{pi_instituition}_{pi}_{self.invoice_month}.pdf", "wb"
+            ) as f:
+                f.write(pdf)
+
         export_data = self._filter_columns()
+
         if not os.path.exists(
             self.name
         ):  # self.name is name of folder storing invoices
             os.mkdir(self.name)
 
         for pi in self.pi_list:
-            _export_pi_invoice(pi)
+            if pandas.isna(pi):
+                continue
+
+            pi_dataframe = self._get_pi_dataframe(export_data, pi)
+            pi_instituition = pi_dataframe[invoice.INSTITUTION_FIELD].iat[0]
+
+            invoice_html_path = _create_html_invoice()
+            _create_pdf_invoice()
+            os.remove(invoice_html_path)
```

Comment:
Instead of manually cleaning up the temporary file, you could make more effective use of `tempfile.NamedTemporaryFile` here by passing an open file into your `_create_*` methods:

```
with tempfile.NamedTemporaryFile(mode='w', suffix='.html') as tempfd:
  _create_html_invoice(tempfd)
  _create_pdf_invoice(tempfd)
```

```
def _create_html_invoice(tempfd):
    environment = Environment(loader=FileSystemLoader(TEMPLATE_DIR_PATH))
    template = environment.get_template("pi_invoice.html")
    content = template.render(
        data=pi_dataframe,
    )

    tempfd.write(content)

    # You might need this to ensure the content is visible in the file.
    tempfd.flush()
```

And you would need similar changes in `_create_pdf_invoice`.

Reply:
@larsks  It turns out I did need to flush it. Thank you!